dev/ledger.md

```diff
@@ -400,73 +403,79 @@ parameters_, which can be encoded as a msgpack struct:
-  value.
+The values in a KV are either
+- `Bytes`, representing a byte-array
+- `Uint`, representing an unsigned 64-bit integer value.
 
-The keys in a TKV are encoded directly as bytes. The maximum length of a key in
-a TKV is 64 bytes. The maximimum length of a `Bytes` value in a TKV is 64 bytes.
+The maximum length of a key in a KV is 64 bytes.
 
 ### State Schemas
 
 A state schema represents limits on the number of each value type that may
-appear in a [TEAL Key/Value Store (TKV)][TEAL Key/Value Stores].
+appear in a [Key/Value Store (KV)][Key/Value Stores]. State schemas
+are used to control the maximum size of global and local state KVs.
 
 A state schema is composed of two fields:
 
-- `NumUint`, encoded as msgpack field `nui`. This field represents the maximum
-  number of integer values that may appear in some TKV.
-- `NumByteSlice`, encoded as msgpack field `nbs`. This field represents the
-  maximum number of byte slice values that may appear in some TKV.
+- `NumUint` represents the maximum number of integer values that may
+  appear in some KV.
+- `NumByteSlice` represents the maximum number of byte-array values
+  that may appear in some KV.
 
-#### App Minimum Balance Increases
+#### App Minimum Balance Changes
 
-When an account opts in to an application or creates an application, the
-minimum balance requirements for that account increase.
+When an account opts in to an application or creates an application,
+the minimum balance requirements for that account increases. The
+minimum balance requirement is decreased equivalently when an account
+closes out or deletes an app.
 
 When opting in to an application, there is a base minimum balance increase
-of 100000 microalgos. There is an additional minimum balance increase based on
+of 100,000 microalgos. There is an additional minimum balance increase based on
 the `LocalStateSchema` for that application, described by following formula:
 
 `28500 * schema.NumUint + 50000 * schema.NumByteSlice` microalgos.
 
 When creating an application, there is a base minimum balance increase
-of 100000 microalgos. There is an additional minimum balance increase
+of 100,000 microalgos. There is an additional minimum balance increase
 of `100000 * ExtraProgramPages` microalgos.  Finally, there is an
 additional minimum balance increase based on the `GlobalStateSchema`
 for that application, described by the following formula:
 
 `28500 * schema.NumUint + 50000 * schema.NumByteSlice` microalgos.
 
+### Boxes
+
+The box store is an associative array mapping keys of type (uint64 x
+byte-array) to values of type byte-array. The key is a pair in which
+the first value corresponds to an existing (or previously existing)
+application ID, and the second is a _box name_, 1 to 64 bytes in
+length.  The value is a byte-array of length not greater than 32,768.
+
+When an application executes an opcode that creates or destroys a box,
```

Comment:
When does this min balance change takes place?
Is it immediate or only at the end of the transaction?
In particular, suppose I create a temporary 32k-box that I then destroy (in the same app call) and I don't have the min balance to pay for that. Would this fail or would this succeed?

Is the `global MinBalance` being updated after each opcode (this would only apply in the weird case where the application account has been rekeyed to a normal account since re-entrancy is forbidden)

Reply:
All min balance checks are done at the very end of a transaction _group_. However, `min_balance` and `acct_params_get MinBalance` report the updated value immediately.

Write tracking includes box creates, though. So if you make big boxes, you might be stopped by your write budget.

If you aren't, and you create and then delete, we don't have to hit the disk (and we don't).